cucumber-expressions/go/cucumber_expression_parser_test.go

```diff
@@ -0,0 +1,344 @@
+package cucumberexpressions
+
+import (
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestCucumberExpressionParser(t *testing.T) {
+	var assertAst = func(t *testing.T, expression string, expected astNode) {
+		ast, err := parse(expression)
+		require.NoError(t, err)
+		require.Equal(t, expected, ast)
+	}
+
+	t.Run("empty string", func(t *testing.T) {
+		assertAst(t, "", astNode{
+			expressionNode,
+			[]astNode{},
+			nullToken,
+		})
+	})
+
+	t.Run("phrase", func(t *testing.T) {
+		assertAst(t, "three blind mice", astNode{
+			expressionNode,
+			[]astNode{
+				{textNode, []astNode{}, token{"three", text}},
+				{textNode, []astNode{}, token{" ", whiteSpace}},
+				{textNode, []astNode{}, token{"blind", text}},
+				{textNode, []astNode{}, token{" ", whiteSpace}},
+				{textNode, []astNode{}, token{"mice", text}},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("optional", func(t *testing.T) {
+		assertAst(t, "(blind)", astNode{
+			expressionNode,
+			[]astNode{
+				{optionalNode,
+					[]astNode{
+						{textNode, []astNode{}, token{"blind", text}},
+					},
+					nullToken,
+				},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("parameter", func(t *testing.T) {
+		assertAst(t, "{string}", astNode{
+			expressionNode,
+			[]astNode{
+				{parameterNode,
+					[]astNode{
+						{textNode, []astNode{}, token{"string", text}},
+					},
+					nullToken,
+				},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("anonymous parameter", func(t *testing.T) {
+		assertAst(t, "{}", astNode{
+			expressionNode,
+			[]astNode{
+				{parameterNode,
+					[]astNode{},
+					nullToken,
+				},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("optional phrase", func(t *testing.T) {
+		assertAst(t, "three (blind) mice", astNode{
+			expressionNode,
+			[]astNode{
+				{textNode, []astNode{}, token{"three", text}},
+				{textNode, []astNode{}, token{" ", whiteSpace}},
+				{optionalNode, []astNode{
+					{textNode, []astNode{}, token{"blind", text}},
+				}, nullToken},
+				{textNode, []astNode{}, token{" ", whiteSpace}},
+				{textNode, []astNode{}, token{"mice", text}},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("slash", func(t *testing.T) {
+		assertAst(t, "\\", astNode{
+			expressionNode,
+			[]astNode{
+				{textNode, []astNode{}, token{"\\", escape}},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("opening brace", func(t *testing.T) {
+		assertAst(t, "{", astNode{
+			expressionNode,
+			[]astNode{
+				{textNode, []astNode{}, token{"{", beginParameter}},
+			},
+			nullToken,
+		})
+	})
+
+	t.Run("unfinished parameter", func(t *testing.T) {
+		assertAst(t, "{string", astNode{
```

Comment:
I am surprised this would not raise an exception instead. Surely if someone meant to `{string` they would escape it: `\\{string`?  (I am genuinely curious as for CucumberExpressions I am implementing I am raising an error for this exact scenario.)

Reply:
Good point.